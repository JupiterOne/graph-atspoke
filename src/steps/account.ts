import {
  createIntegrationEntity,
  IntegrationStep,
  IntegrationStepExecutionContext,
} from '@jupiterone/integration-sdk-core';

import { IntegrationConfig } from '../types';
import { createAPIClient } from '../client';

export const DATA_ACCOUNT_ENTITY = 'DATA_ACCOUNT_ENTITY';

export async function fetchAccountDetails({
  instance,
  jobState,
}: IntegrationStepExecutionContext<IntegrationConfig>) {
  const apiClient = createAPIClient(instance.config);
  const acctInfo = await apiClient.getAccountInfo();
  const name = `atSpoke ${acctInfo.org} - ${instance.name}`;
  const accountEntity = await jobState.addEntity(
    createIntegrationEntity({
      entityData: {
        source: {
          id: 'atSpoke',
          name: 'atSpoke Account',
        },
        assign: {
          _key: `at-spoke-account:${instance.id}`,
          _type: 'atspoke_account',
          _class: 'Account',
          name,
          displayName: name,
          org: acctInfo.org,
        },
      },
    }),
  );

  await jobState.setData(DATA_ACCOUNT_ENTITY, accountEntity);
}

export const accountSteps: IntegrationStep<IntegrationConfig>[] = [
  {
    id: 'fetch-account',
    name: 'Fetch Account Details',
    entities: [
      {
        resourceName: 'atSpoke Account',
        _type: 'atspoke_account',
        _class: 'Account',
      },
    ],
    relationships: [],
    dependsOn: [],
    executionHandler: fetchAccountDetails,
  },
];
