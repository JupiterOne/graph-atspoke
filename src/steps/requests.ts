import {
  createDirectRelationship,
  createIntegrationEntity,
  Entity,
  IntegrationStep,
  IntegrationStepExecutionContext,
  RelationshipClass,
} from '@jupiterone/integration-sdk-core';

import { createAPIClient } from '../client';
import { IntegrationConfig } from '../types';
import { DATA_ACCOUNT_ENTITY } from './account';

export async function fetchRequests({
  instance,
  jobState,
}: IntegrationStepExecutionContext<IntegrationConfig>) {
  const apiClient = createAPIClient(instance.config);

  const accountEntity = (await jobState.getData(DATA_ACCOUNT_ENTITY)) as Entity;

  await apiClient.iterateRequests(async (request) => {
    if (request.requestTypeInfo) {
      delete request.requestTypeInfo;
    }
    const requestEntity = await jobState.addEntity(
      createIntegrationEntity({
        entityData: {
          source: request,
          assign: {
            _type: 'at_spoke_request',
            _class: 'Record',
            _key: request.id,
            partial: true,
            name: request.subject,
            displayName: request.subject,
            webLink: request.permalink,
            email: request.email,
            status: request.status,
            requester: request.requester,
            owner: request.owner,
            requestType: request.requestType,
          },
        },
      }),
    );

    await jobState.addRelationship(
      createDirectRelationship({
        _class: RelationshipClass.HAS,
        from: accountEntity,
        to: requestEntity,
      }),
    );

    if (request.requestType) {
      const requestTypeEntity = await jobState.findEntity(request.requestType);
      if (requestTypeEntity) {
        await jobState.addRelationship(
          createDirectRelationship({
            _class: RelationshipClass.HAS,
            from: requestEntity,
            to: requestTypeEntity,
          }),
        );
      }
    }
  });
}

export async function fetchRequestTypes({
  instance,
  jobState,
}: IntegrationStepExecutionContext<IntegrationConfig>) {
  const apiClient = createAPIClient(instance.config);

  const accountEntity = (await jobState.getData(DATA_ACCOUNT_ENTITY)) as Entity;

  await apiClient.iterateRequestTypes(async (requestType) => {
    const requestTypeEntity = await jobState.addEntity(
      createIntegrationEntity({
        entityData: {
          source: requestType,
          assign: {
            _type: 'at_spoke_requesttype',
            _class: 'Configuration',
            _key: requestType.id,
            partial: true,
            name: requestType.title,
            displayName: requestType.title,
            description: requestType.description,
            status: requestType.status,
          },
        },
      }),
    );

    await jobState.addRelationship(
      createDirectRelationship({
        _class: RelationshipClass.HAS,
        from: accountEntity,
        to: requestTypeEntity,
      }),
    );
  });
}

export const requestSteps: IntegrationStep<IntegrationConfig>[] = [
  {
    id: 'fetch-requests',
    name: 'Fetch Requests',
    entities: [
      {
        resourceName: 'atSpoke Request',
        _type: 'at_spoke_request',
        _class: 'Record',
      },
    ],
    relationships: [
      {
        _type: 'at_spoke_account_has_request',
        _class: RelationshipClass.HAS,
        sourceType: 'at_spoke_account',
        targetType: 'at_spoke_request',
      },
      {
        _type: 'at_spoke_request_has_requesttype',
        _class: RelationshipClass.HAS,
        sourceType: 'at_spoke_request',
        targetType: 'at_spoke_requesttype',
      },
    ],
    dependsOn: ['fetch-request-types'],
    executionHandler: fetchRequests,
  },
  {
    id: 'fetch-request-types',
    name: 'Fetch Request Types',
    entities: [
      {
        resourceName: 'atSpoke Request Type',
        _type: 'at_spoke_requesttype',
        _class: 'Configuration',
      },
    ],
    relationships: [
      {
        _type: 'at_spoke_account_has_requesttype',
        _class: RelationshipClass.HAS,
        sourceType: 'at_spoke_account',
        targetType: 'at_spoke_requesttype',
      },
    ],
    dependsOn: ['fetch-account'],
    executionHandler: fetchRequestTypes,
  },
];
