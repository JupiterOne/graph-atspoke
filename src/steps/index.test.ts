import {
  createMockStepExecutionContext,
  Recording,
} from '@jupiterone/integration-sdk-testing';

import { IntegrationConfig } from '../types';
import { setupSpokeRecording } from '../../test/recording';
import { fetchTeams, fetchUsers } from './access';
import { fetchAccountDetails } from './account';
import { fetchRequests } from './requests';
import { fetchWebhooks } from './webhooks';
import { stopFetchingRequests } from '../client';
import { AtSpokeRequest } from '../client';

const DEFAULT_API_KEY = 'fake_api_key'; // works because we have a recording now

const integrationConfig: IntegrationConfig = {
  apiKey: process.env.API_KEY || DEFAULT_API_KEY,
};

jest.setTimeout(1000 * 60 * 1);

let recording: Recording;

afterEach(async () => {
  await recording.stop();
});

test('should collect data', async () => {
  recording = setupSpokeRecording({
    directory: __dirname,
    name: 'steps',
    redactedRequestHeaders: ['api-key'],
  });

  const context = createMockStepExecutionContext<IntegrationConfig>({
    instanceConfig: integrationConfig,
  });

  // Simulates dependency graph execution.
  // See https://github.com/JupiterOne/sdk/issues/262.
  await fetchAccountDetails(context);
  await fetchUsers(context);
  await fetchTeams(context);
  await fetchWebhooks(context);
  await fetchRequests(context);
  //because fetchRequests only goes back 14 days, you might expect this fetch
  //to fail on the old recording in the test suite, but fetchRequests always
  //pulls the first page of requests regardless, which is enough for the
  //half dozen requests in the test recording

  // Review snapshot, failure is a regression
  expect({
    numCollectedEntities: context.jobState.collectedEntities.length,
    numCollectedRelationships: context.jobState.collectedRelationships.length,
    collectedEntities: context.jobState.collectedEntities,
    collectedRelationships: context.jobState.collectedRelationships,
    encounteredTypes: context.jobState.encounteredTypes,
  }).toMatchSnapshot();

  const accounts = context.jobState.collectedEntities.filter((e) =>
    e._class.includes('Account'),
  );
  expect(accounts.length).toBeGreaterThan(0);
  expect(accounts).toMatchGraphObjectSchema({
    _class: ['Account'],
    schema: {
      additionalProperties: true,
      properties: {
        _type: { const: 'atspoke_account' },
        manager: { type: 'string' },
        _rawData: {
          type: 'array',
          items: { type: 'object' },
        },
      },
      required: ['org'], //we use this to make webLinks to users
    },
  });

  const users = context.jobState.collectedEntities.filter((e) =>
    e._class.includes('User'),
  );
  expect(users.length).toBeGreaterThan(0);
  expect(users).toMatchGraphObjectSchema({
    _class: ['User'],
    schema: {
      additionalProperties: true,
      properties: {
        _type: { const: 'atspoke_user' },
        firstName: { type: 'string' },
        _rawData: {
          type: 'array',
          items: { type: 'object' },
        },
      },
      required: ['email'], //we use this to make webLinks and even names if name is blank
    },
  });

  const userGroups = context.jobState.collectedEntities.filter((e) =>
    e._class.includes('UserGroup'),
  );
  expect(userGroups.length).toBeGreaterThan(0);
  expect(userGroups).toMatchGraphObjectSchema({
    _class: ['UserGroup'],
    schema: {
      additionalProperties: true,
      properties: {
        _type: { const: 'atspoke_team' },
        _rawData: {
          type: 'array',
          items: { type: 'object' },
        },
      },
      required: [],
    },
  });

  //webhooks and requests won't exist on all accts
  const webhooks = context.jobState.collectedEntities.filter((e) =>
    e._class.includes('ApplicationEndpoint'),
  );
  expect(webhooks.length).toBeGreaterThan(0);
  expect(webhooks).toMatchGraphObjectSchema({
    _class: ['ApplicationEndpoint'],
    schema: {
      additionalProperties: true,
      properties: {
        _type: { const: 'atspoke_webhook' },
        _rawData: {
          type: 'array',
          items: { type: 'object' },
        },
      },
      required: [],
    },
  });
});

describe('stopFetchingRequests', () => {
  const requests: AtSpokeRequest[] = [];
  let lastExecutionTime = 0; //beginning of epoch
  const templateRequest: AtSpokeRequest = {
    subject: 'Tough stuff',
    requester: 'Somebody',
    owner: 'Nobody',
    status: 'OPEN',
    privacyLevel: 'high',
    team: 'winners',
    org: 'Team Co.',
    permalink: 'not applicable',
    id: '333333333333',
    isAutoResolve: false,
    isFiled: true,
    email: 'help@help.com',
    createdAt: new Date().toDateString(), //today, right now
    updatedAt: new Date().toDateString(), //today, right now
  };
  function getTimeDaysAgo(daysAgo: number): number {
    return new Date().getTime() - daysAgo * 24 * 60 * 60 * 1000;
  }
  function getDaysAgoDateString(daysAgo: number): string {
    return new Date(getTimeDaysAgo(daysAgo)).toDateString();
  }

  test('no requests in page', () => {
    expect(stopFetchingRequests(requests, lastExecutionTime)).toBe(true);
  });

  test('lastExecutionTime 0 ingests max age', () => {
    templateRequest.updatedAt = getDaysAgoDateString(13);
    requests.push(templateRequest);
    lastExecutionTime = 0;
    //last request is not older than 14 days or exTime, so do not stop
    expect(stopFetchingRequests(requests, lastExecutionTime)).toBe(false);
    templateRequest.updatedAt = getDaysAgoDateString(15);
    requests.push(templateRequest);
    //last request is older than 14 days, so stop
    expect(stopFetchingRequests(requests, lastExecutionTime)).toBe(true);
  });

  test('lastRequest missing updateAt stops ingestion', () => {
    templateRequest.updatedAt = '';
    requests.push(templateRequest);
    lastExecutionTime = 0;
    expect(stopFetchingRequests(requests, lastExecutionTime)).toBe(true);
  });

  test('lastExecutionTime 2 days, pagination stops on time', () => {
    templateRequest.updatedAt = getDaysAgoDateString(1);
    requests.push(templateRequest);
    lastExecutionTime = getTimeDaysAgo(2);
    //last request is not older than 14 days or lastExecutionTime, so do not stop
    expect(stopFetchingRequests(requests, lastExecutionTime)).toBe(false);
    templateRequest.updatedAt = getDaysAgoDateString(3);
    requests.push(templateRequest);
    //last request is older than lastExecutionTime, so stop
    expect(stopFetchingRequests(requests, lastExecutionTime)).toBe(true);
  });

  test('Date() unparsable string for lastRequest updatedAt stops ingestion', () => {
    templateRequest.updatedAt = 'JupiterOne is the best';
    requests.push(templateRequest);
    lastExecutionTime = 0;
    expect(stopFetchingRequests(requests, lastExecutionTime)).toBe(true);
  });
});
