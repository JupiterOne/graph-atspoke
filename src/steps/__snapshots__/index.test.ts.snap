// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should collect data 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Account",
      ],
      "_key": "at-spoke-account:local-integration-instance",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "id": "atSpoke",
            "name": "atSpoke Account",
          },
        },
      ],
      "_type": "at_spoke_account",
      "createdOn": undefined,
      "displayName": "atSpoke jupiteronedev - Local Integration",
      "id": "atSpoke",
      "name": "atSpoke jupiteronedev - Local Integration",
      "org": "jupiteronedev",
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "602bebeeb9f1b80007da131c",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "createdAt": "2021-02-16T15:59:42.436Z",
            "displayName": "",
            "email": "adam.williams@jupiterone.com",
            "id": "602bebeeb9f1b80007da131c",
            "isEmailVerified": false,
            "isProfileCompleted": false,
            "memberships": Array [
              Object {
                "org": "jupiteronedev",
                "role": "MEMBER",
                "status": "INVITED",
              },
            ],
            "profile": Object {
              "countryCode": "1",
              "countryId": "US",
              "formattedPhoneNumber": "",
              "isManager": false,
              "locale": "en-US",
            },
          },
        },
      ],
      "_type": "at_spoke_user",
      "createdOn": 1613491182436,
      "displayName": "adam.williams@jupiterone.com",
      "email": "adam.williams@jupiterone.com",
      "id": "602bebeeb9f1b80007da131c",
      "isEmailVerified": false,
      "isProfileCompleted": false,
      "name": "adam.williams@jupiterone.com",
      "status": undefined,
      "username": "adam.williams@jupiterone.com",
      "webLink": "https://jupiteronedev.askspoke.com/users/602bebeeb9f1b80007da131c",
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "602585518843b90007eeda34",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "createdAt": "2021-02-11T19:28:17.296Z",
            "displayName": "",
            "email": "keitoshi.casey@gmail.com",
            "id": "602585518843b90007eeda34",
            "isEmailVerified": true,
            "isProfileCompleted": false,
            "memberships": Array [
              Object {
                "org": "jupiteronedev",
                "role": "MEMBER",
                "status": "INVITED",
              },
            ],
            "profile": Object {
              "countryCode": "1",
              "countryId": "US",
              "formattedPhoneNumber": "",
              "isManager": false,
              "locale": "en-US",
            },
          },
        },
      ],
      "_type": "at_spoke_user",
      "createdOn": 1613071697296,
      "displayName": "keitoshi.casey@gmail.com",
      "email": "keitoshi.casey@gmail.com",
      "id": "602585518843b90007eeda34",
      "isEmailVerified": true,
      "isProfileCompleted": false,
      "name": "keitoshi.casey@gmail.com",
      "status": undefined,
      "username": "keitoshi.casey@gmail.com",
      "webLink": "https://jupiteronedev.askspoke.com/users/602585518843b90007eeda34",
    },
    Object {
      "_class": Array [
        "User",
      ],
      "_key": "601da2951d25d80007bf87d2",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "createdAt": "2021-02-05T19:55:01.769Z",
            "displayName": "Kevin Casey",
            "email": "thomaskevincasey74@gmail.com",
            "id": "601da2951d25d80007bf87d2",
            "isEmailVerified": true,
            "isProfileCompleted": true,
            "memberships": Array [
              Object {
                "org": "jupiteronedev",
                "role": "ADMIN",
                "status": "ACTIVE",
              },
            ],
            "profile": Object {
              "countryCode": "1",
              "countryId": "US",
              "formattedPhoneNumber": "",
              "isManager": false,
              "locale": "en-US",
            },
          },
        },
      ],
      "_type": "at_spoke_user",
      "createdOn": 1612554901769,
      "displayName": "Kevin Casey",
      "email": "thomaskevincasey74@gmail.com",
      "id": "601da2951d25d80007bf87d2",
      "isEmailVerified": true,
      "isProfileCompleted": true,
      "name": "Kevin Casey",
      "status": undefined,
      "username": "Kevin Casey",
      "webLink": "https://jupiteronedev.askspoke.com/users/601da2951d25d80007bf87d2",
    },
    Object {
      "_class": Array [
        "UserGroup",
      ],
      "_key": "601da2f204e2850006039574",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "_id": "601da2f204e2850006039574",
            "analyticsDashboardId": "d871625f78fc",
            "canRoute": true,
            "cdcPartId": "61df53b2-99d7-4e39-8de4-c57522932888",
            "cdcTransactionId": "2956890b-ab22-4111-b194-35df68eebecd",
            "color": "blue",
            "createdAt": "2021-02-05T19:56:34.846Z",
            "createdFromSample": false,
            "description": "The first test team",
            "email": "test-team-alpha@jupiteronedev.askspoke.com",
            "emailChannel": Object {
              "emailForwards": Array [
                Object {
                  "_id": "601da2f204e2850006039576",
                  "createdAt": "2021-02-05T19:56:34.852Z",
                  "origin": "test-team-alpha@jupiteronedev.askspoke.com",
                  "recipient": "test-team-alpha@jupiteronedev.askspoke.com",
                },
              ],
              "id": null,
            },
            "icon": "support",
            "iconPng": "https://static.askspoke.com/img/icons/teams/png/icon-teams-support-blue.png",
            "iconPngLarge": "https://static.askspoke.com/img/icons/teams/png/icon-teams-support-blue@2x.png",
            "id": "601da2f204e2850006039574",
            "keywords": Array [],
            "name": "Test Team Alpha",
            "org": "jupiteronedev",
            "owner": "601da2951d25d80007bf87d2",
            "permalink": "https://jupiteronedev.askspoke.com/teams/test-team-alpha",
            "settings": Object {
              "analytics": Object {
                "customReports": Array [],
              },
              "delegation": Object {
                "excludedUsers": Array [],
                "strategy": "ROUND_ROBIN",
              },
              "flags": Object {
                "AUTO_DELEGATION": true,
                "AUTO_PRIVACY": false,
                "INVITE_ONLY": true,
              },
            },
            "slug": "test-team-alpha",
            "status": "ACTIVE",
            "updatedAt": "2021-02-05T19:56:34.846Z",
          },
        },
      ],
      "_type": "at_spoke_team",
      "active": false,
      "createdOn": 1612554994846,
      "description": "The first test team",
      "displayName": "Test Team Alpha",
      "email": "test-team-alpha@jupiteronedev.askspoke.com",
      "id": "601da2f204e2850006039574",
      "name": "Test Team Alpha",
      "org": "jupiteronedev",
      "owner": "601da2951d25d80007bf87d2",
      "status": "ACTIVE",
      "webLink": "https://jupiteronedev.askspoke.com/teams/test-team-alpha",
    },
    Object {
      "_class": Array [
        "UserGroup",
      ],
      "_key": "601da2991d25d80007bf8852",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "_id": "601da2991d25d80007bf8852",
            "analyticsDashboardId": "d871625f78fc",
            "canRoute": true,
            "cdcPartId": "b1c776c8-cf45-4020-98ec-28f02fa931fe",
            "cdcTransactionId": "0b839b4b-abad-4b88-971c-d24d09d7bc01",
            "color": "green",
            "createdAt": "2021-02-05T19:55:05.757Z",
            "createdFromSample": false,
            "description": "Tech support and services.",
            "email": "it@jupiteronedev.askspoke.com",
            "emailChannel": Object {
              "emailForwards": Array [
                Object {
                  "_id": "601da2991d25d80007bf8854",
                  "createdAt": "2021-02-05T19:55:05.761Z",
                  "origin": "it@jupiteronedev.askspoke.com",
                  "recipient": "it@jupiteronedev.askspoke.com",
                },
              ],
              "id": null,
            },
            "goals": Object {
              "FIRST_REPLY": Object {
                "createdAt": "2021-02-05T19:55:05.760Z",
                "durationInSec": 86400,
                "enabled": true,
                "recipient": "AGENTS",
                "updatedAt": "2021-02-05T19:55:06.715Z",
              },
            },
            "icon": "it",
            "iconPng": "https://static.askspoke.com/img/icons/teams/png/icon-teams-it-green.png",
            "iconPngLarge": "https://static.askspoke.com/img/icons/teams/png/icon-teams-it-green@2x.png",
            "id": "601da2991d25d80007bf8852",
            "keywords": Array [],
            "name": "IT",
            "org": "jupiteronedev",
            "owner": "601da2951d25d80007bf87d2",
            "permalink": "https://jupiteronedev.askspoke.com/teams/it",
            "settings": Object {
              "analytics": Object {
                "customReports": Array [],
              },
              "delegation": Object {
                "excludedUsers": Array [],
                "strategy": "ROUND_ROBIN",
              },
              "flags": Object {
                "AUTO_DELEGATION": true,
                "AUTO_PRIVACY": false,
                "INVITE_ONLY": true,
              },
            },
            "slug": "it",
            "status": "ACTIVE",
            "teamGoalLabel": "1 day",
            "teamType": "it",
            "updatedAt": "2021-02-05T19:55:05.757Z",
          },
        },
      ],
      "_type": "at_spoke_team",
      "active": false,
      "createdOn": 1612554905757,
      "description": "Tech support and services.",
      "displayName": "IT",
      "email": "it@jupiteronedev.askspoke.com",
      "id": "601da2991d25d80007bf8852",
      "name": "IT",
      "org": "jupiteronedev",
      "owner": "601da2951d25d80007bf87d2",
      "status": "ACTIVE",
      "webLink": "https://jupiteronedev.askspoke.com/teams/it",
    },
    Object {
      "_class": Array [
        "ApplicationEndpoint",
      ],
      "_key": "6026fcda2ae4920006bb8661",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "authentication": Object {
              "clientSecret": "REDACTED_SECRET",
              "type": "client_secret",
            },
            "client": "Zapier",
            "createdAt": "2021-02-12T22:10:34.105Z",
            "description": "Push resource creation to external system.",
            "enabled": true,
            "id": "6026fcda2ae4920006bb8661",
            "org": "jupiteronedev",
            "topics": Array [
              "request.created",
            ],
            "url": "https://example.com/receive_hook",
            "user": "601da2951d25d80007bf87d2",
          },
        },
      ],
      "_type": "at_spoke_webhook",
      "address": "https://example.com/receive_hook",
      "createdOn": 1613167834105,
      "description": "Push resource creation to external system.",
      "displayName": "Zapier",
      "enabled": true,
      "id": "6026fcda2ae4920006bb8661",
      "name": "Zapier",
      "targetServiceName": "Zapier",
      "targetUrl": "https://example.com/receive_hook",
      "topics": Array [
        "request.created",
      ],
    },
    Object {
      "_class": Array [
        "Record",
      ],
      "_key": "6026f691ae0b320006140f3e",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "createdAt": "2021-02-12T21:43:45.630Z",
            "email": "request-6026f691ae0b320006140f3e@jupiteronedev.askspoke.com",
            "id": "6026f691ae0b320006140f3e",
            "isAutoResolved": false,
            "isFiled": true,
            "org": "jupiteronedev",
            "owner": "602585518843b90007eeda34",
            "permalink": "https://jupiteronedev.askspoke.com/next/requests/4",
            "privacyLevel": "public",
            "requestType": "6026f670ac67a90007a282f1",
            "requester": "601da2951d25d80007bf87d2",
            "status": "OPEN",
            "subject": "Request for “Broken hardware”",
            "tags": Array [],
            "team": "601da2f204e2850006039574",
            "updatedAt": "2021-02-12T21:43:45.902Z",
          },
        },
      ],
      "_type": "at_spoke_request",
      "active": false,
      "createdOn": 1613166225630,
      "displayName": "Request for “Broken hardware”",
      "email": "request-6026f691ae0b320006140f3e@jupiteronedev.askspoke.com",
      "name": "Request for “Broken hardware”",
      "owner": "602585518843b90007eeda34",
      "partial": true,
      "requestType": "6026f670ac67a90007a282f1",
      "requester": "601da2951d25d80007bf87d2",
      "status": "OPEN",
      "webLink": "https://jupiteronedev.askspoke.com/next/requests/4",
    },
    Object {
      "_class": Array [
        "Record",
      ],
      "_key": "6026d93a0a4bad0007bf52ee",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "createdAt": "2021-02-12T19:39:08.071Z",
            "email": "request-6026d93a0a4bad0007bf52ee@jupiteronedev.askspoke.com",
            "id": "6026d93a0a4bad0007bf52ee",
            "isAutoResolved": false,
            "isFiled": true,
            "org": "jupiteronedev",
            "owner": "602585518843b90007eeda34",
            "permalink": "https://jupiteronedev.askspoke.com/next/requests/3",
            "privacyLevel": "public",
            "requester": "601da2951d25d80007bf87d2",
            "status": "OPEN",
            "subject": "Test msg for Slack integration",
            "tags": Array [],
            "team": "601da2f204e2850006039574",
            "updatedAt": "2021-02-12T19:39:08.452Z",
          },
        },
      ],
      "_type": "at_spoke_request",
      "active": false,
      "createdOn": 1613158748071,
      "displayName": "Test msg for Slack integration",
      "email": "request-6026d93a0a4bad0007bf52ee@jupiteronedev.askspoke.com",
      "name": "Test msg for Slack integration",
      "owner": "602585518843b90007eeda34",
      "partial": true,
      "requestType": undefined,
      "requester": "601da2951d25d80007bf87d2",
      "status": "OPEN",
      "webLink": "https://jupiteronedev.askspoke.com/next/requests/3",
    },
    Object {
      "_class": Array [
        "Record",
      ],
      "_key": "602585518843b90007eeda39",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "createdAt": "2021-02-11T19:28:17.420Z",
            "email": "request-602585518843b90007eeda39@jupiteronedev.askspoke.com",
            "id": "602585518843b90007eeda39",
            "isAutoResolved": false,
            "isFiled": true,
            "lastResolvedAt": "2021-02-12T19:15:05.540Z",
            "org": "jupiteronedev",
            "owner": "601da2951d25d80007bf87d2",
            "permalink": "https://jupiteronedev.askspoke.com/next/requests/1",
            "privacyLevel": "public",
            "requester": "602585518843b90007eeda34",
            "resolveTime": 85608120,
            "status": "RESOLVED",
            "subject": "Testing email to see if it works",
            "tags": Array [],
            "team": "601da2f204e2850006039574",
            "teamResponseTimeInMs": 415586,
            "updatedAt": "2021-02-12T19:15:05.793Z",
          },
        },
      ],
      "_type": "at_spoke_request",
      "active": false,
      "createdOn": 1613071697420,
      "displayName": "Testing email to see if it works",
      "email": "request-602585518843b90007eeda39@jupiteronedev.askspoke.com",
      "name": "Testing email to see if it works",
      "owner": "601da2951d25d80007bf87d2",
      "partial": true,
      "requestType": undefined,
      "requester": "602585518843b90007eeda34",
      "status": "RESOLVED",
      "webLink": "https://jupiteronedev.askspoke.com/next/requests/1",
    },
    Object {
      "_class": Array [
        "Record",
      ],
      "_key": "6025858242c00500073e2f6f",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": Object {
            "createdAt": "2021-02-11T19:29:06.567Z",
            "email": "request-6025858242c00500073e2f6f@jupiteronedev.askspoke.com",
            "id": "6025858242c00500073e2f6f",
            "isAutoResolved": false,
            "isFiled": true,
            "org": "jupiteronedev",
            "permalink": "https://jupiteronedev.askspoke.com/next/requests/2",
            "privacyLevel": "public",
            "requester": "602585518843b90007eeda34",
            "status": "OPEN",
            "subject": "Sample request - where is the printer?",
            "tags": Array [],
            "updatedAt": "2021-02-11T19:29:06.586Z",
          },
        },
      ],
      "_type": "at_spoke_request",
      "active": false,
      "createdOn": 1613071746567,
      "displayName": "Sample request - where is the printer?",
      "email": "request-6025858242c00500073e2f6f@jupiteronedev.askspoke.com",
      "name": "Sample request - where is the printer?",
      "owner": undefined,
      "partial": true,
      "requestType": undefined,
      "requester": "602585518843b90007eeda34",
      "status": "OPEN",
      "webLink": "https://jupiteronedev.askspoke.com/next/requests/2",
    },
  ],
  "collectedRelationships": Array [
    Object {
      "_class": "HAS",
      "_fromEntityKey": "at-spoke-account:local-integration-instance",
      "_key": "at-spoke-account:local-integration-instance|has|602bebeeb9f1b80007da131c",
      "_toEntityKey": "602bebeeb9f1b80007da131c",
      "_type": "at_spoke_account_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "at-spoke-account:local-integration-instance",
      "_key": "at-spoke-account:local-integration-instance|has|602585518843b90007eeda34",
      "_toEntityKey": "602585518843b90007eeda34",
      "_type": "at_spoke_account_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "at-spoke-account:local-integration-instance",
      "_key": "at-spoke-account:local-integration-instance|has|601da2951d25d80007bf87d2",
      "_toEntityKey": "601da2951d25d80007bf87d2",
      "_type": "at_spoke_account_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "at-spoke-account:local-integration-instance",
      "_key": "at-spoke-account:local-integration-instance|has|601da2f204e2850006039574",
      "_toEntityKey": "601da2f204e2850006039574",
      "_type": "at_spoke_account_has_team",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "601da2f204e2850006039574",
      "_key": "601da2f204e2850006039574|has|601da2951d25d80007bf87d2",
      "_toEntityKey": "601da2951d25d80007bf87d2",
      "_type": "at_spoke_team_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "601da2f204e2850006039574",
      "_key": "601da2f204e2850006039574|has|602585518843b90007eeda34",
      "_toEntityKey": "602585518843b90007eeda34",
      "_type": "at_spoke_team_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "601da2f204e2850006039574",
      "_key": "601da2f204e2850006039574|has|602bebeeb9f1b80007da131c",
      "_toEntityKey": "602bebeeb9f1b80007da131c",
      "_type": "at_spoke_team_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "at-spoke-account:local-integration-instance",
      "_key": "at-spoke-account:local-integration-instance|has|601da2991d25d80007bf8852",
      "_toEntityKey": "601da2991d25d80007bf8852",
      "_type": "at_spoke_account_has_team",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "601da2991d25d80007bf8852",
      "_key": "601da2991d25d80007bf8852|has|601da2951d25d80007bf87d2",
      "_toEntityKey": "601da2951d25d80007bf87d2",
      "_type": "at_spoke_team_has_user",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "at-spoke-account:local-integration-instance",
      "_key": "at-spoke-account:local-integration-instance|has|6026fcda2ae4920006bb8661",
      "_toEntityKey": "6026fcda2ae4920006bb8661",
      "_type": "at_spoke_account_has_webhook",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "at-spoke-account:local-integration-instance",
      "_key": "at-spoke-account:local-integration-instance|has|6026f691ae0b320006140f3e",
      "_toEntityKey": "6026f691ae0b320006140f3e",
      "_type": "at_spoke_account_has_request",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "at-spoke-account:local-integration-instance",
      "_key": "at-spoke-account:local-integration-instance|has|6026d93a0a4bad0007bf52ee",
      "_toEntityKey": "6026d93a0a4bad0007bf52ee",
      "_type": "at_spoke_account_has_request",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "at-spoke-account:local-integration-instance",
      "_key": "at-spoke-account:local-integration-instance|has|602585518843b90007eeda39",
      "_toEntityKey": "602585518843b90007eeda39",
      "_type": "at_spoke_account_has_request",
      "displayName": "HAS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "at-spoke-account:local-integration-instance",
      "_key": "at-spoke-account:local-integration-instance|has|6025858242c00500073e2f6f",
      "_toEntityKey": "6025858242c00500073e2f6f",
      "_type": "at_spoke_account_has_request",
      "displayName": "HAS",
    },
  ],
  "encounteredTypes": Array [
    "at_spoke_account",
    "at_spoke_user",
    "at_spoke_account_has_user",
    "at_spoke_team",
    "at_spoke_account_has_team",
    "at_spoke_team_has_user",
    "at_spoke_webhook",
    "at_spoke_account_has_webhook",
    "at_spoke_request",
    "at_spoke_account_has_request",
  ],
  "numCollectedEntities": 11,
  "numCollectedRelationships": 14,
}
`;
